In [30]:
import pandas as pd

df = pd.read_csv("data/Telco-Customer-Churn.csv")

print("Số dòng, số cột:", df.shape)
display(df.head())
df.info()

Số dòng, số cột: (7043, 21)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [31]:
# Đổi TotalCharges sang số; ô không chuyển được sẽ thành NaN
total_charges_numeric = pd.to_numeric(df["TotalCharges"], errors="coerce")

print("Số giá trị TotalCharges không chuyển được sang số:")
print(total_charges_numeric.isna().sum())

print("\nCác dòng cần kiểm tra:")
display(df.loc[total_charges_numeric.isna(),
               ["customerID", "tenure", "MonthlyCharges", "TotalCharges", "Churn"]])

Số giá trị TotalCharges không chuyển được sang số:
11

Các dòng cần kiểm tra:


,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


In [32]:
# Tạo bản để xử lý, giữ nguyên df đọc từ file gốc
df_clean = df.copy()

# Chuyển TotalCharges từ chữ sang số
df_clean["TotalCharges"] = pd.to_numeric(
    df_clean["TotalCharges"], errors="coerce"
)

# Chỉ điền 0 cho các dòng vừa thiếu TotalCharges vừa có tenure = 0
mask = df_clean["TotalCharges"].isna() & (df_clean["tenure"] == 0)
df_clean.loc[mask, "TotalCharges"] = 0

print("Số dòng được điền 0:", mask.sum())
print("TotalCharges còn thiếu:", df_clean["TotalCharges"].isna().sum())
print("Kiểu dữ liệu mới:", df_clean["TotalCharges"].dtype)

Số dòng được điền 0: 11
TotalCharges còn thiếu: 0
Kiểu dữ liệu mới: float64


In [33]:
print("Số dòng, số cột:", df_clean.shape)
print("customerID bị trùng:", df_clean["customerID"].duplicated().sum())
print("Số dòng trùng hoàn toàn:", df_clean.duplicated().sum())

print("\nSố giá trị thiếu theo từng cột:")
print(df_clean.isna().sum().to_string())

Số dòng, số cột: (7043, 21)
customerID bị trùng: 0
Số dòng trùng hoàn toàn: 0

Số giá trị thiếu theo từng cột:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0


In [34]:
print("Các giá trị trong cột Churn:")
print(df_clean["Churn"].value_counts())

total_customers = len(df_clean)
churned_customers = (df_clean["Churn"] == "Yes").sum()
churn_rate = churned_customers / total_customers * 100

print(f"\nTổng khách hàng: {total_customers:,}")
print(f"Khách hàng rời bỏ: {churned_customers:,}")
print(f"Tỷ lệ churn: {churn_rate:.2f}%")

Các giá trị trong cột Churn:
Churn
No     5174
Yes    1869
Name: count, dtype: int64

Tổng khách hàng: 7,043
Khách hàng rời bỏ: 1,869
Tỷ lệ churn: 26.54%


In [35]:
churn_by_contract = (
    df_clean.groupby("Contract")
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda x: (x == "Yes").sum())
    )
    .reset_index()
)

churn_by_contract["churn_rate"] = (
    churn_by_contract["churned_customers"]
    / churn_by_contract["total_customers"]
    * 100
).round(2)

display(churn_by_contract)

,Contract,total_customers,churned_customers,churn_rate
0,Month-to-month,3875,1655,42.71
1,One year,1473,166,11.27
2,Two year,1695,48,2.83


Khách dùng hợp đồng theo tháng có tỷ lệ churn 42,71%, cao hơn nhiều so với nhóm hợp đồng một năm (11,27%) và hai năm (2,83%). Trong 1.869 khách rời bỏ, có 1.655 người thuộc nhóm hợp đồng theo tháng. Đây là nhóm nên được ưu tiên tìm hiểu để đề xuất chương trình giữ chân

In [37]:
# Chia khách hàng theo số tháng đã sử dụng dịch vụ
df_clean["tenure_group"] = pd.cut(
    df_clean["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0–12 tháng", "13–24 tháng", "25–48 tháng", "49–72 tháng"]
)

# Tính số khách và tỷ lệ rời bỏ trong từng nhóm
churn_by_tenure = (
    df_clean.groupby("tenure_group", observed=True)
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

churn_by_tenure["churn_rate"] = (
    churn_by_tenure["churned_customers"]
    / churn_by_tenure["total_customers"] * 100
).round(2)

display(churn_by_tenure)

,tenure_group,total_customers,churned_customers,churn_rate
0,0–12 tháng,2186,1037,47.44
1,13–24 tháng,1024,294,28.71
2,25–48 tháng,1594,325,20.39
3,49–72 tháng,2239,213,9.51


Nhóm 0–12 tháng có tỷ lệ churn cao nhất, 47,44%; họ cũng chiếm 1.037 trong 1.869 khách rời bỏ. Đây là nhóm đáng ưu tiên xem xét trong kế hoạch giữ chân khách hàng. Lưu ý đây là so sánh giữa các nhóm trong một bộ dữ liệu, chưa chứng minh rằng cứ sử dụng lâu thêm thì nguy cơ rời bỏ của cùng một khách hàng sẽ giảm.

In [39]:
contract_tenure = (
    df_clean.groupby(["tenure_group", "Contract"], observed=True)
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

contract_tenure["churn_rate"] = (
    contract_tenure["churned_customers"]
    / contract_tenure["total_customers"] * 100
).round(2)

display(contract_tenure)

,tenure_group,Contract,total_customers,churned_customers,churn_rate
0,0–12 tháng,Month-to-month,1994,1024,51.35
1,0–12 tháng,One year,124,13,10.48
2,0–12 tháng,Two year,68,0,0.00
3,13–24 tháng,Month-to-month,737,278,37.72
4,13–24 tháng,One year,197,16,8.12
5,13–24 tháng,Two year,90,0,0.00
6,25–48 tháng,Month-to-month,802,264,32.92
7,25–48 tháng,One year,518,55,10.62
8,25–48 tháng,Two year,274,6,2.19
9,49–72 tháng,Month-to-month,342,89,26.02


Trong từng nhóm thời gian sử dụng, khách dùng hợp đồng theo tháng đều có tỷ lệ churn cao hơn khách dùng hợp đồng một năm hoặc hai năm. Nhóm 0–12 tháng dùng hợp đồng theo tháng ghi nhận 1.024 khách rời bỏ trên 1.994 khách, tương đương 51,35%. Đây là phân khúc nên được ưu tiên tìm hiểu khi đề xuất biện pháp giữ chân. Kết quả cho thấy mối liên hệ trong dữ liệu, chưa chứng minh loại hợp đồng là nguyên nhân gây churn.

In [41]:
churn_by_internet = (
    df_clean.groupby("InternetService")
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

churn_by_internet["churn_rate"] = (
    churn_by_internet["churned_customers"]
    / churn_by_internet["total_customers"] * 100
).round(2)

display(churn_by_internet)

,InternetService,total_customers,churned_customers,churn_rate
0,DSL,2421,459,18.96
1,Fiber optic,3096,1297,41.89
2,No,1526,113,7.40


Khách dùng Fiber optic có tỷ lệ churn 41,89%, cao nhất trong ba nhóm dịch vụ Internet. Nhóm này ghi nhận 1.297 khách rời bỏ. Cần xem thêm loại hợp đồng của họ trước khi đề xuất biện pháp giữ chân.

In [43]:
internet_contract = (
    df_clean.groupby(["Contract", "InternetService"])
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

internet_contract["churn_rate"] = (
    internet_contract["churned_customers"]
    / internet_contract["total_customers"] * 100
).round(2)

display(internet_contract)

,Contract,InternetService,total_customers,churned_customers,churn_rate
0,Month-to-month,DSL,1223,394,32.22
1,Month-to-month,Fiber optic,2128,1162,54.61
2,Month-to-month,No,524,99,18.89
3,One year,DSL,570,53,9.30
4,One year,Fiber optic,539,104,19.29
5,One year,No,364,9,2.47
6,Two year,DSL,628,12,1.91
7,Two year,Fiber optic,429,31,7.23
8,Two year,No,638,5,0.78


Trong từng loại hợp đồng, khách dùng Fiber optic đều có tỷ lệ churn cao hơn khách dùng DSL. Nhóm Fiber optic dùng hợp đồng theo tháng ghi nhận 1.162/2.128 khách rời bỏ, tương đương 54,61%. Đây là phân khúc cần được ưu tiên tìm hiểu thêm; dữ liệu hiện tại chưa xác định được nguyên nhân rời bỏ.

In [45]:
df_clean["monthly_charge_group"] = pd.cut(
    df_clean["MonthlyCharges"],
    bins=[0, 35, 65, 95, float("inf")],
    labels=["≤35", ">35–65", ">65–95", ">95"]
)

churn_by_charge = (
    df_clean.groupby("monthly_charge_group", observed=True)
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

churn_by_charge["churn_rate"] = (
    churn_by_charge["churned_customers"]
    / churn_by_charge["total_customers"] * 100
).round(2)

display(churn_by_charge)

,monthly_charge_group,total_customers,churned_customers,churn_rate
0,≤35,1735,189,10.89
1,>35–65,1409,326,23.14
2,>65–95,2604,936,35.94
3,>95,1295,418,32.28


Nhóm có MonthlyCharges từ trên 65 đến 95 ghi nhận tỷ lệ churn cao nhất (35,94%) và 936 khách rời bỏ. Tỷ lệ ở nhóm trên 95 là 32,28%, nên dữ liệu không cho thấy churn tăng đều theo mức phí. Cần xét thêm loại dịch vụ Internet trước khi diễn giải chênh lệch này.

In [47]:
internet_charge = (
    df_clean.groupby(
        ["InternetService", "monthly_charge_group"],
        observed=True
    )
    .agg(
        total_customers=("customerID", "count"),
        churned_customers=("Churn", lambda values: (values == "Yes").sum())
    )
    .reset_index()
)

internet_charge["churn_rate"] = (
    internet_charge["churned_customers"]
    / internet_charge["total_customers"] * 100
).round(2)

display(internet_charge)

,InternetService,monthly_charge_group,total_customers,churned_customers,churn_rate
0,DSL,≤35,209,76,36.36
1,DSL,>35–65,1409,326,23.14
2,DSL,>65–95,803,57,7.10
3,Fiber optic,>65–95,1801,879,48.81
4,Fiber optic,>95,1295,418,32.28
5,No,≤35,1526,113,7.40


Tỷ lệ churn chung là 26,54%. Các phân khúc nổi bật gồm khách sử dụng 0–12 tháng, khách dùng hợp đồng theo tháng và khách dùng Fiber optic. Nhóm Fiber optic kết hợp hợp đồng theo tháng ghi nhận 1.162 khách rời bỏ trên 2.128 khách (54,61%). Trong khoảng phí >65–95, tỷ lệ churn của Fiber optic (48,81%) cao hơn DSL (7,10%). Đây là các mối liên hệ quan sát được trong dữ liệu; cần tìm hiểu thêm trước khi kết luận nguyên nhân hoặc đưa ra ưu đãi cụ thể.

In [49]:
output_path = "data/telco_customer_churn_clean.csv"

df_clean.to_csv(output_path, index=False, encoding="utf-8-sig")

print("Đã lưu:", output_path)
print("Kích thước bảng:", df_clean.shape)
print("TotalCharges còn thiếu:", df_clean["TotalCharges"].isna().sum())

Đã lưu: data/telco_customer_churn_clean.csv
Kích thước bảng: (7043, 23)
TotalCharges còn thiếu: 0
